### PR TITLE
[examples] add pyhf example notebook

### PR DESCRIPTION
**Pull request by lorenzennio**
Add a pyhf example (see #807).

TODOs:
- [x] fix renaming through `parameter_map`. This is not actually renaming the parameter, but only the qualified name.
- [ ] fix warning ```RuntimeWarning: invalid value encountered in subtract
  return xlogy(n, lam) - lam - gammaln(n + 1.0)  # type: ignore[no-any-return]```

### CODE REVIEW

**Comment by lorenzennio** [comment]
@dvandyk any updates on how to assign a custom observable as a pyhf parameter here?

File: python/eos/pyhf_likelihood.py [hunk 1]
```diff
@@ -44,31 +44,52 @@ def __init__(self, cache, workspace, parameter_map={}):
...
+                    options = eos.Options(parameter_map[parameter_name].get('options', {}))
+                else:
+                    raise ValueError('parameter_map values must be either strings or dictionaries.')
             else:
-                qn = eos.QualifiedName('pyhf::' + pn)
+                qualified_name = eos.QualifiedName('pyhf::' + parameter_name)
 
-            if self.parameters.has(qn):
-                p = self.parameters[qn]
+            if self.parameters.has(qualified_name):
```
Comment (dvandyk) [inline python/eos/pyhf_likelihood.py:73]: The logic here is off. You have 3 cases:
1. ``qualified_name`` corresponds to an existing parameter;
2. ``qualified_name`` corresponds to an observable name;
3. ``qualified_name`` is unknown to EOS.

You are currently handling case 1 in line 74, and both cases 2 and 3 in lines 76...83. That is definitely buggy. I am unsure if that causes the warning you see, though.

Comment (lorenzennio) [reply python/eos/pyhf_likelihood.py]: Ah I see. This indeed solves the problem. Before it created a new parameter with the same name as the observable, even if it existed. Hence, the warning. Should be fixed now.

File: python/eos/pyhf_likelihood.py [hunk 2]
```diff
@@ -44,31 +44,60 @@ def __init__(self, cache, workspace, parameter_map={}):
...
+                else:
+                    raise ValueError('parameter_map values must be either strings or dictionaries.')
             else:
-                qn = eos.QualifiedName('pyhf::' + pn)
+                qualified_name = eos.QualifiedName('pyhf::' + parameter_name)
 
-            if self.parameters.has(qn):
-                p = self.parameters[qn]
+            # check if qualified name corresponds to an existing observable
+            if qualified_name in [o[0] for o in eos.Observables()]:
```
Comment (lorenzennio) [inline python/eos/pyhf_likelihood.py:74]: Is there a better way to check if an observable exists?

File: python/eos/pyhf_likelihood.py [hunk 3]
```diff
@@ -79,7 +108,12 @@ def factory(cache, workspace, parameter_map={}):
     def priors(self):
         priors = []
         for name, _ in self.model.config.modifiers:
-            param_name = self.parameter_map[name] if name in self.parameter_map else 'pyhf::' + name
+            param_name = 'pyhf::' + name
+            if name in self.parameter_map:
+                if isinstance(self.parameter_map[name], str):
+                    param_name = self.parameter_map[name]
+                elif isinstance(self.parameter_map[name], dict):
+                    continue
```
Comment (lorenzennio) [inline python/eos/pyhf_likelihood.py:116]: I assume now that if a dictionary is passed, it is an observable and hence no prior is assigned. It is the desired behaviour, but maybe not the best way to implement it.

File: python/eos/pyhf_likelihood.py (hunk 1 above)
Comment (dvandyk) [reply python/eos/pyhf_likelihood.py]: Brilliant!

File: python/eos/pyhf_likelihood.py (hunk 2 above)
Comment (lorenzennio) [reply python/eos/pyhf_likelihood.py]: Done!

**Comment by dvandyk** [comment]
I still would like to go over this with you at some point prior to merging. For example, I am puzzled by the fact that the distribution in C_VL^ublnu is flat. Shouldn't you be able to extract C_VL in the 2nd fit?

**Comment by lorenzennio** [comment]
> I still would like to go over this with you at some point prior to merging. For example, I am puzzled by the fact that the distribution in C_VL^ublnu is flat. Shouldn't you be able to extract C_VL in the 2nd fit?

Ah, I am still working on the notebook and what to include. So far the pure pyhf posterior looks like this: 

image

**Comment by lorenzennio** [comment]
@dvandyk now the notebook is updated too. Let me know what you think (:

**Comment by lorenzennio** [comment]
@dvandyk  from my side we can mergre this

**Comment by lorenzennio** [comment]
@dvandyk I would be happy to merge this, as I would like to show this in a bayesian pyhf workshop soon. 

There is still one warning, which comes from the fact that the Gaussian and Poisson priors in eos are unbounded, whereas pyhf parameters are bounded. But this is not too much of an issue if your model is sensible. We could create a separate issue for this and fix it later.

**Review by dvandyk — APPROVED** [review]
Thank you @lorenzennio, merging this ASAP.

**Pull request merged**